### ISSUE TITLE
Distribution error: Tensor of shape..

### ISSUE THREAD

**Issue by sirdawar** [StatMixedML/XGBoostLSS#35]
First of all thanks for the new pytorch version.
I've been using the previous versions and today saw a new version and wanted to give it a try, with the data and code that worked fine in previous version.

After I edited the old code with the differences of the new version, following the examples, i've noticed some problems with distributions.

```
So when I give my label data to the optimization or training I get something on the lines of
Expected value argument (Tensor of shape (8700,)) to be within the support (GreaterThan(lower_bound=0.0)) of the distribution LogNormal(), but found invalid values:
tensor([0.2782, 0.3064, 0.3202,  ..., 0.3338, 0.3202, 0.3202])

or

Expected parameter scale (Tensor of shape ()) of distribution Normal(loc: -974.14453125, scale: -1210.6866455078125) to satisfy the constraint GreaterThan(lower_bound=0.0), but found invalid values:
-1210.6866455078125
```
This happens with every distribution that I've tried. 
My dataset is between 0.09 and 0.9 in value, I've tried with similar datasets and got similar results. With one dataset I've managed to run the model by multiplying the values by 10, but for other datasets does not work.

Reminding that all these datasets worked fine with previous version. Do you know what might be the reason?

**Comment by StatMixedML** [comment]
@sirdawar Thanks for the interest in the project.

> My dataset is between 0.09 and 0.9 in value, ....

This seems like a `Beta` distributed dataset, so fitting a continuous real distribution like the `Normal` can cause issues. 

But since you`ve tried other distributions also, this might not be the reason. I suspect that it has to do with the starting values. I look into it and report back.

**Comment by maxfield-green** [comment]
Hi there, I believe I have potentially ran into the same issue of predicted distribution shape. I've been using this library for a few months, with the "Expectile" distribution. After the upgrade to pytorch, I noticed a dramatic shift in the scale and location of predicted data distributions. It seems that test predictions are being re-centered at zero, and also the expectile flip in orientation i.e. 0.95 with lower valued distribution that 0.05.

I made some adjustments to synthetic gaussian simulation example and found the same case to be true.

<img width="572" alt="image" src="https://github.com/StatMixedML/XGBoostLSS/assets/126024321/c1f85c70-6193-4fc8-90fc-195787d0536a">

<img width="572" alt="image" src="https://github.com/StatMixedML/XGBoostLSS/assets/126024321/109d9e30-4027-4000-8e3d-b7c2bfb08357">

**Comment by StatMixedML** [comment]
@maxfield-green Thanks for your good analysis.

Would you be so kind to share the notebook, so that I can look into and replicate the issue.

Thanks!

**Comment by StatMixedML** [comment]
@maxfield-green Ok cool, thanks for the notebook. 

Setting [base_margin](https://github.com/StatMixedML/XGBoostLSS/blob/master/xgboostlss/model.py#L115) is not something you do, but it is implicit in the code. So no worries, nothing you can account for. 

@sirdawar, @maxfield-green I report back once your problems are fixed.

**Comment by StatMixedML** [comment]
@sirdawar, @maxfield-green I have made changes to the corresponding functions. Can you please re-install the package and test if it is working. Thanks.

**Comment by sirdawar** [comment]
Just tried again using your distribution select example and also the optimization example. And still getting similar errors. 
I attached one example of X_train and Y_train data that you can check. 

```
Expected parameter loc (Tensor of shape (1, 1)) of distribution Normal(loc: tensor([[nan]], grad_fn=), scale: tensor([[nan]], grad_fn=)) to satisfy the constraint Real(), but found invalid values:
tensor([[nan]], grad_fn=)
```
[Traindata_for xgblss_test.zip](https://github.com/StatMixedML/XGBoostLSS/files/11565616/Traindata_for.xgblss_test.zip)

**Comment by StatMixedML** [comment]
@sirdawar I have made changes to the corresponding functions. Can you please re-install the package and test if it is working now. Also, please use the following notebook

**Comment by sirdawar** [comment]
That works now for choosing distribution, but still can't make to run hyperparameter tuning.
```

Expected parameter concentration (Tensor of shape (13850, 1, 2)) of distribution Dirichlet(concentration: torch.Size([13850, 1, 2])) to satisfy the constraint IndependentConstraint(GreaterThan(lower_bound=0.0), 1), but found invalid values:
tensor([[[0.0302,    nan]],

        [[0.7349,    nan]],

        [[2.3822,    nan]],

        ...,

        [[1.7256,    nan]],

        [[0.0510,    nan]],

        [[0.0316,    nan]]])
```

**Comment by StatMixedML** [comment]
@sirdawar Can you re-install the package and try again with the following notebook.
[Beta Example.zip](https://github.com/StatMixedML/XGBoostLSS/files/11572524/Beta.Example.zip)

**Comment by sirdawar** [comment]
This now works, I tried Beta, LogNormal and Gaussian.

Beta and LogNormal works fine, with Gaussian it runs optimization, but then when running predict it gives again the error:
```
Expected parameter loc (Tensor of shape (4210,)) of distribution Normal(loc: torch.Size([4210]), scale: torch.Size([4210])) to satisfy the constraint Real(), but found invalid values:
tensor([nan, nan, nan,  ..., nan, nan, nan])
```

**Comment by StatMixedML** [comment]
@sirdawar Glad that the `Beta` and `LogNormal` are now working.

Given the Skewness and the support `(0,1)` of the data, I wouldn't recommend using a symmetric distribution like the `Gaussian`, since this very likely leads to numerical instabilities. This is also what you see from your output. 

In case you still want to use the `Gaussian` you can try and
- change the response function using `softplus`
- transform the response via. e.g., `y*10` and see if this is running

I have attached an example that runs end-to-end using the `Gaussian`. I used the `softplus` instead of `exp` when initializing the distribution. 

Thanks again for using the package!

[Gaussian Example.zip](https://github.com/StatMixedML/XGBoostLSS/files/11575225/Gaussian.Example.zip)

**Comment by StatMixedML** [comment]
@maxfield-green Do you still have the problems when estimating expectiles?

**Comment by StatMixedML** [comment]
@sirdawar The previous version mostly used analytical gradients and hessians. 

With the new release v0.2.0 we now rely exclusively on PyTorch and its automatic derivation of gradients and hessians. Automatic differentiation enables efficient computation of gradients and hessians and also offers greater flexibility to  incorporate custom loss functions into XGBoostLSS workflows. Hence, it is easier to add new distributions and also have a more unified structure among distributions.

As far as varying results are concerned: there are some differences between scipy and PyTorch distributions, i.e., how distributions are parameterized. Yet, we shouldn't expect to see a lot of differences.

**Comment by StatMixedML** [comment]
Closing this issue.
